src/main/java/emu/grasscutter/game/quest/GameMainQuest.java

```diff
@@ -125,9 +125,36 @@ public GameQuest getChildQuestByOrder(int order) {
     }
 
     public void finish() {
+        // Avoid recursion from child finish() in GameQuest
+        // when auto finishing all child quests with QUEST_STATE_UNFINISHED (below)
+        if (this.isFinished) {
+            Grasscutter.getLogger().debug("Skip main quest finishing because it's already finished");
+            return;
+        }
+
         this.isFinished = true;
         this.state = ParentQuestState.PARENT_QUEST_STATE_FINISHED;
 
+         /*
+            We also need to check for unfinished quests in the main quest line,
```

Comment:
just a little nitpick: `main quest line` can easily be interpreted as `quest series` (i.e. series of parent quests), which is totally different from what you mean here. Otherwise looks good
```suggestion
            We also need to check for unfinished childQuests in this MainQuest
```